### PR TITLE
feat(endpoints): Add support for global config 

### PR DESCRIPTION
**Pull request by TBS1996**
wip

### CODE REVIEW

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 1]
```diff
@@ -27,6 +27,10 @@ def _sample_apm():
     return random.random() < getattr(settings, "SENTRY_RELAY_ENDPOINT_APM_SAMPLING", 0)
 
 
+def get_global_config():
```
Comment (iker-barriocanal) [inline src/sentry/api/endpoints/relay/project_configs.py:30]: This is part of the config computation already. What do you think about moving it to the `sentry.relay` module with the rest of the project config functionality?

**Review by iker-barriocanal — COMMENTED** [review]
This first version looks promising!

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 2]
```diff
@@ -49,9 +53,13 @@ def _post(self, request: Request):
             return Response("Relay unauthorized for full config information", 403)
 
         version = request.GET.get("version") or "1"
+        global_config: bool = request.GET.get("global")
```
[the remark this answers is not in the document]
Comment (TBS1996) [reply src/sentry/api/endpoints/relay/project_configs.py]: good idea! I changed it to your 'global_requested' suggestion

File: src/sentry/api/endpoints/relay/project_configs.py (hunk 1 above)
Comment (TBS1996) [reply src/sentry/api/endpoints/relay/project_configs.py]: i made it simply take the values from the project config computation, I thought we could avoid getting those values if global config isn't requested to save on some processing, but at least now we save on network traffic and on relay memory.  
 
whats your thoughts on that? my thinking now was that, we can do it this simple way first, and then later try to be more specific with what the `_get_cached_or_schedule` function returns later

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 3]
```diff
@@ -49,13 +49,43 @@ def _post(self, request: Request):
...
+            # global config.
+            res = self._post_or_schedule_by_key(request)
+            metrics.incr("relay.project_configs.post_v4.pending", amount=len(res["pending"]))
+            metrics.incr("relay.project_configs.post_v4.fetched", amount=len(res["configs"]))
+
+            measurements = res["measurements"]
+            metricsConditionalTagging = res["metricsConditionalTagging"]
+
+            if global_requested:
+                global_config = {
```
Comment (TBS1996) [inline src/sentry/api/endpoints/relay/project_configs.py:67]: should i make a metric here about the requesting of a global config?

File: src/sentry/api/endpoints/relay/project_configs.py (hunk 1 above)
Comment (iker-barriocanal) [reply src/sentry/api/endpoints/relay/project_configs.py]: I'm not sure if I understand. My suggestion was to move the function into another module with the rest of the functions, not to reuse the same function itself. What do you think about separating these?

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 4]
```diff
@@ -49,13 +49,35 @@ def _post(self, request: Request):
             return Response("Relay unauthorized for full config information", 403)
 
         version = request.GET.get("version") or "1"
+
         set_tag("relay_protocol_version", version)
 
-        if self._should_use_v3(version, request):
+        if version == "4":
+            # With v4, we want to move 'measurements' and 'metricsConditionalTagging' into a new
+            # global config.
```
Comment (iker-barriocanal) [inline src/sentry/api/endpoints/relay/project_configs.py:57]: This comment will become stale when we modify the config, what do you think if we remove it?

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 5]
```diff
@@ -49,13 +49,35 @@ def _post(self, request: Request):
...
+
+            del res["measurements"]
+            del res["metricsConditionalTagging"]
+
+            return Response(res, status=200)
+        elif self._should_use_v3(version, request):
             # Always compute the full config. It's invalid to send partial
             # configs to processing relays, and these validate the requests they
             # get with permissions and trim configs down accordingly.
-            return self._post_or_schedule_by_key(request)
```
Comment (iker-barriocanal) [inline src/sentry/api/endpoints/relay/project_configs.py:58]: The code for `v3` was easier to read before this change. What do you think if we restore it, and extract common functionality into a separate function instead?

Comment (TBS1996) [reply src/sentry/api/endpoints/relay/project_configs.py:58]: I'm a bit confused, the common functionality is already extracted, I removed the two lines specific to 'v3' in '_post_or_schedule_by_key' and reused that function for v4

do you mean encapsulating the extra logic for v3 and v4 into two new functions (and those then calling '_post_or_schedule_by_key' ?

Comment (iker-barriocanal) [reply src/sentry/api/endpoints/relay/project_configs.py:58]: Yep, the second case sounds like a good step toward easier readability!

Comment (TBS1996) [reply src/sentry/api/endpoints/relay/project_configs.py:58]: ok, i made it so now

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 6]
```diff
@@ -49,13 +49,13 @@ def _post(self, request: Request):
...
+
         set_tag("relay_protocol_version", version)
 
-        if self._should_use_v3(version, request):
-            # Always compute the full config. It's invalid to send partial
-            # configs to processing relays, and these validate the requests they
-            # get with permissions and trim configs down accordingly.
-            return self._post_or_schedule_by_key(request)
+        if version == "4":
+            return self._get_v4_config
```
Comment (iker-barriocanal) [inline src/sentry/api/endpoints/relay/project_configs.py:56]: Should we call the function here?

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 7]
```diff
@@ -67,7 +67,33 @@ def _post(self, request: Request):
...
+    def _get_v4_config(self, request):
+        res = self._post_or_schedule_by_key(request)
+        metrics.incr("relay.project_configs.post_v4.pending", amount=len(res["pending"]))
+        metrics.incr("relay.project_configs.post_v4.fetched", amount=len(res["configs"]))
+
+        if request.GET.get("global"):
+            res["global"] = {
+                "measurements": res["measurements"],
+                "metricsConditionalTagging": res["metricsConditionalTagging"],
+            }
```
Comment (iker-barriocanal) [inline src/sentry/api/endpoints/relay/project_configs.py:81]: What do you think about having a separate function for the global config, and merge it? I think something like the following could work:
```suggestion
        if request.GET.get("global"):
            res["global"] = _get_global_config(...)

# ...

def _get_global_config():
  return {
    "measurements": get_measurements_config(...),
    "metricsConditionalTagging": get_metrics_conditional_tagging_rules(...)
  }
```

Another benefit of isolating the global config computation is that it's easier to test it. What do you think?

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 8]
```diff
@@ -67,7 +67,33 @@ def _post(self, request: Request):
                 full_config_requested=full_config_requested,
             )
         else:
-            return Response("Unsupported version, we only support versions 1 to 3.", 400)
+            return Response("Unsupported version, we only support versions 1 to 4.", 400)
+
+    def _get_v4_config(self, request):
+        res = self._post_or_schedule_by_key(request)
+        metrics.incr("relay.project_configs.post_v4.pending", amount=len(res["pending"]))
+        metrics.incr("relay.project_configs.post_v4.fetched", amount=len(res["configs"]))
```
Comment (iker-barriocanal) [inline src/sentry/api/endpoints/relay/project_configs.py:75]: In the current state, it's possible a request comes in with no projects but requesting the global config. In this case, emitting a counter metric with `amount=0` doesn't provide any observability insights but still creates the overhead of generating the metric. What do you think if we only emit these when `amount>0`?

Comment (iker-barriocanal) [reply src/sentry/api/endpoints/relay/project_configs.py:75]: Related -- is there any observability we can add that could help us get more insights around global configs, the new functionality in `v4`?

File: src/sentry/api/endpoints/relay/project_configs.py (hunk 6 above)
Comment (TBS1996) [reply src/sentry/api/endpoints/relay/project_configs.py:56]: lol shit 🤦‍♂️ absolutely, nice catch

File: src/sentry/api/endpoints/relay/project_configs.py (hunk 8 above)
Comment (TBS1996) [reply src/sentry/api/endpoints/relay/project_configs.py:76]: sounds good to me, I suppose we would wanna do it on v3 too?

for the 2nd, I suppose you mean adding a similar metric to global right? that makes a lot of sense to me

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 9]
```diff
@@ -43,19 +43,21 @@ def _post(self, request: Request):
         relay = request.relay
         assert relay is not None  # should be provided during Authentication
 
-        full_config_requested = request.relay_request_data.get("fullConfig")
+        if request.relay_request_data.get("globalConfig"):
```
Comment (TBS1996) [inline src/sentry/api/endpoints/relay/project_configs.py:46]: iiuc, `global_config` gets turned into camelcase `globalConfig`

File: src/sentry/api/endpoints/relay/project_configs.py [hunk 10]
```diff
@@ -43,19 +43,21 @@ def _post(self, request: Request):
...
+            metrics.incr("relay.project_configs.global.fetched")
```
Comment (TBS1996) [inline src/sentry/api/endpoints/relay/project_configs.py:47]: can i just use a new metric like this or do i have to declare it somewhere else?

File: src/sentry/api/endpoints/relay/project_configs.py (hunk 8 above)
Comment (TBS1996) [reply src/sentry/api/endpoints/relay/project_configs.py:77]: now we won't request project configs along with globalconfig anymore, as we request globalconfig totally separately, so then i dont think we need to change this

**Pull request closed without merging**